plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/DockerConnector.java

```diff
@@ -1720,4 +1656,51 @@ private void addQueryParamIfNotNull(DockerConnection connection, String queryPar
         }
     }
 
+    // TODO add comments to each line :)
+
+    // todo check when closed by interruption exception is thrown
+
+    // Here do some trick to be able interrupt build process. Basically for now it is not possible interrupt docker daemon while
+    // it's building images but here we need just be able to close connection to the unix socket. Thread is blocking while read
+    // from the socket stream so need one more thread that is able to close socket. In this way we can release thread that is
+    // blocking on i/o.
+
+    /**
+     * This method runs provided runnable in a separate thread in a such a way that that thread can be interrupted.
+     * It is supposed that it runs long-running task and at some point it can be stopped.
+     * <br/>Provided task should throw unchecked exception if exceptional situation happens.
+     * In that case {@link IOException#IOException(String, Throwable)} will be thrown
+     * with {@link Throwable#getLocalizedMessage()} and unchecked exception as arguments.
+     */
+    private void executeInSeparateThread(Runnable task) throws InterruptedException, IOException {
+        final ValueHolder<Throwable> errorHolder = new ValueHolder<>();
```

Comment:
What about to use 
https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html
instead of this method?